### PR TITLE
example variant walker

### PR DESCRIPTION
**Pull request by akiezun**

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/tools/examples/ExampleVariantWalker.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        oneLineSummary = "Print variants with optional contextual data",
+        programGroup = VariantProgramGroup.class
+)
+public final class ExampleVariantWalker extends VariantWalker {
+
+    @Argument(fullName = StandardArgumentDefinitions.OUTPUT_LONG_NAME, shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME, doc = "Output file (if not provided, defaults to STDOUT)", common = false, optional = true)
+    public File outputFile = null;
+
+    @Argument(fullName="auxiliaryVariants", shortName="av", doc="Auxiliary set of variants", optional=true)
+    public FeatureInput<Feature> mask;
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/examples/ExampleVariantWalker.java]: If these are supposed to be variants, should this be a `FeatureInput<VariantContext>`?

File: src/main/java/org/broadinstitute/hellbender/tools/examples/ExampleVariantWalker.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+    private PrintStream outputStream = null;
+
+    @Override
+    public void onTraversalStart() {
+        super.onTraversalStart();
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/examples/ExampleVariantWalker.java]: It is not necessary/required to call `super.onTraversalStart()` here -- that is the whole reason we separated `super.onTraversalStart()` from `onStartup()`

File: src/main/java/org/broadinstitute/hellbender/tools/examples/ExampleVariantWalker.java (hunk 1 above)
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/examples/ExampleVariantWalker.java]: Also, the name `mask` is confusing/puzzling -- how about `auxiliaryVariants`?

File: src/test/java/org/broadinstitute/hellbender/tools/CountVariantsIntegrationTest.java [hunk 3]
```diff
@@ -9,6 +10,11 @@
 
 public final class CountVariantsIntegrationTest extends CommandLineProgramTest {
 
+    @Override
+    public String getTestedClassName() {
+        return CountVariants.class.getSimpleName();
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/CountVariantsIntegrationTest.java]: Isn't this what the superclass version already does? Why the override?

File: src/test/java/org/broadinstitute/hellbender/tools/examples/ExampleVariantWalkerIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+import org.testng.annotations.Test;
+
+import java.io.IOException;
+import java.util.Arrays;
+
+public final class ExampleVariantWalkerIntegrationTest extends CommandLineProgramTest {
+    @Override
+    public String getTestedClassName() {
+        return ExampleVariantWalker.class.getSimpleName();
+    }
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/examples/ExampleVariantWalkerIntegrationTest.java]: Here too -- doesn't the superclass version of `getTestedClassName()` already do this?

File: src/test/resources/org/broadinstitute/hellbender/engine/example_variants.vcf [hunk 5]
```diff
@@ -0,0 +1,14 @@
+##fileformat=VCFv4.1
+#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO
```
Comment (droazen) [inline src/test/resources/org/broadinstitute/hellbender/engine/example_variants.vcf]: Erm, this is just a slightly modified version of `feature_data_source_test.vcf`, which you are using as the auxiliary input to your test. At least change the locations and alleles....

**Comment by droazen** [comment]
Review complete -- back to @akiezun

File: src/test/resources/org/broadinstitute/hellbender/tools/expected_ExampleVariantWalkerIntegrationTest_output.txt [hunk 6]
```diff
@@ -0,0 +1,22 @@
+Current variant: [VC Unknown @ 1:100 Q1.00 of type=SNP alleles=[G*, A] attr={} GT=[]
```
Comment (droazen) [inline src/test/resources/org/broadinstitute/hellbender/tools/expected_ExampleVariantWalkerIntegrationTest_output.txt]: Should be in `tools/examples`?

File: src/main/java/org/broadinstitute/hellbender/tools/examples/ExampleVariantWalker.java (hunk 1 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/examples/ExampleVariantWalker.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/examples/ExampleVariantWalker.java (hunk 2 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/examples/ExampleVariantWalker.java]: this is copy-pasted from `ExampleIntervalWalker`. Fixed now in both

File: src/test/java/org/broadinstitute/hellbender/tools/CountVariantsIntegrationTest.java (hunk 3 above)
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/tools/CountVariantsIntegrationTest.java]: the override is to make the IDE know that the class `CountVariants` is used.

File: src/test/java/org/broadinstitute/hellbender/tools/examples/ExampleVariantWalkerIntegrationTest.java (hunk 4 above)
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/tools/examples/ExampleVariantWalkerIntegrationTest.java]: same reason

File: src/test/resources/org/broadinstitute/hellbender/engine/example_variants.vcf (hunk 5 above)
Comment (akiezun) [inline src/test/resources/org/broadinstitute/hellbender/engine/example_variants.vcf]: done

File: src/test/resources/org/broadinstitute/hellbender/tools/expected_ExampleVariantWalkerIntegrationTest_output.txt (hunk 6 above)
Comment (akiezun) [inline src/test/resources/org/broadinstitute/hellbender/tools/expected_ExampleVariantWalkerIntegrationTest_output.txt]: another copy paste from `ExampleIntervalWalker`. Fixed in both

**Comment by akiezun** [comment]
back to @droazen

File: src/test/java/org/broadinstitute/hellbender/tools/examples/ExampleVariantWalkerIntegrationTest.java (hunk 4 above)
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/examples/ExampleVariantWalkerIntegrationTest.java]: The problem with this is that this class is an example of how to write tests for a tool -- it's bad to have unnecessary things like this in examples.

File: src/main/java/org/broadinstitute/hellbender/tools/examples/ExampleVariantWalker.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+        oneLineSummary = "Print variants with optional contextual data",
+        programGroup = VariantProgramGroup.class
+)
+public final class ExampleVariantWalker extends VariantWalker {
+
+    @Argument(fullName = StandardArgumentDefinitions.OUTPUT_LONG_NAME, shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME, doc = "Output file (if not provided, defaults to STDOUT)", common = false, optional = true)
+    public File outputFile = null;
+
+    @Argument(fullName="auxiliaryVariants", shortName="av", doc="Auxiliary set of variants", optional=true)
+    public FeatureInput<VariantContext> auxiliaryVariants;
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/examples/ExampleVariantWalker.java]: Make `@Argument` fields `private` -- we don't want people to mistakenly think that they have to be `public` in order for the arg injection to work (here and above in `ExampleIntervalWalker`)

**Comment by droazen** [comment]
Two more comments to address, then feel free to merge.

File: src/test/java/org/broadinstitute/hellbender/tools/examples/ExampleVariantWalkerIntegrationTest.java (hunk 4 above)
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/tools/examples/ExampleVariantWalkerIntegrationTest.java]: hmm, not sure. I like to be able to quickly navigate from the test to the tested class, which this reference enables. Without this reference I need to open a special dialog to type in the class name etc. Plus, the class then looks unused to the IDE.

File: src/main/java/org/broadinstitute/hellbender/tools/examples/ExampleVariantWalker.java (hunk 7 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/examples/ExampleVariantWalker.java]: fine, but if they're private they look unused in the IDE which bothers me

File: src/test/java/org/broadinstitute/hellbender/tools/examples/ExampleVariantWalkerIntegrationTest.java (hunk 4 above)
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/examples/ExampleVariantWalkerIntegrationTest.java]: It's an example of how to write a test class though -- we definitely should not have unnecessary and confusing stuff like this here!

Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/tools/examples/ExampleVariantWalkerIntegrationTest.java]: i think it's anti-confusing actually - it helps. will delete this for you though

Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/examples/ExampleVariantWalkerIntegrationTest.java]: Overriding a superclass method to do exactly what it already does is "anti-confusing"? :)

**Comment by droazen** [comment]
You broke `FeatureSupportIntegrationTest` by renaming `PrintReadsWithVariants`...

**Comment by akiezun** [comment]
fixed #323

**Pull request merged**